### ISSUE TITLE
BUG: ndimage.center_of_mass yields different results Windows vs. Linux

### ISSUE THREAD

**Issue by RPGillespie6** [scipy/scipy#16821]
### Describe your issue.

`ndimage.center_of_mass` yeilds different results on the same input when run from Windows vs. Linux.

- Both scipy versions are `1.9.0`
- Both numpy versions are `1.23.1`
- Both python versions are `3.8`

When the code snippet is run under Windows 10, it produces the following result (incorrect result):
`(-9095.696809099283, -8984.458504552724, -5286.132827965857)`

When the code snippet is run under Ubuntu 20.04, it produces the following result (correct result):
`(255.1366754713629, 252.01641192226484, 148.27740899169206)`

A copy of the input (compressed) can be downloaded here: https://drive.google.com/file/d/1WdFr9YJea-YQBpDM-52ow3R0jiBUO5Yw/view?usp=sharing

### Reproducing Code Example

```python
import numpy as np
from scipy import ndimage
arr = np.load('./arr.npy')
print(ndimage.center_of_mass(arr)) # Prints different results Windows 10 vs. Ubuntu 20.04, all other things equal
```

### Error message

```shell
n/a
```

### SciPy/NumPy/Python version information

1.9.0 1.23.1 sys.version_info(major=3, minor=8, micro=3, releaselevel='final', serial=0)

**Comment by tupui** [comment]
Hi @RPGillespie6, thank you for reporting. Could you reproduce this with a minimal example? I am afraid I had to remove your link for security reasons. We don't want to download archived files. An alternative is to post a gist with plain text.

**Comment by RPGillespie6** [comment]
This issue was discovered while calculating center of mass of CT scan image slices from anonymized DICOM, hence why the input is a large compressed file. It might take me a long time to distill it down to a minimal example. What is the security concern, exactly? It's just a pickled blob of numerical data (`.npy`).

**Comment by tupui** [comment]
I am afraid it is going to be hard to help without a small reproducer.

The security concern is that, pardon me for saying this, we don't know each other and I don't know how I can trust a compressed file you would send me. Anyone can post anything on GitHub and it's very easy to craft malformed/corrupt files. Even if we were to know each, I could go rogue and be an accomplice of something harmful. I am not sure for GitHub itself, but there are precedents of all that all around (e.g. in the NPM space with maintainers breaking their own popular packages).

**Comment by RPGillespie6** [comment]
It's not an executable file. So you don't have to "trust" it. It's just a blob of data. 

>  it's very easy to craft malformed/corrupt files

It is, but malformed/corrupt files are not dangerous, just broken. Still not understanding what you are afraid of, exactly. Are you afraid there is a vulnerability in numpy that will be exploited upon loading the data? 

My suggestion is:
1. Download my file in Linux first using wget or curl - it's just a file and can't hurt you unless you attempt executing it
2. Unpickle in linux and verify the file is just plain old data
3. Repickle the data yourself for use on Windows

**Comment by tupui** [comment]
Have a look at tar/zip bomb. This is just a simple example of what can be done. There are way more advanced things than that. Just attempting to open a malicious compressed file can compromise you.

**Comment by RPGillespie6** [comment]
Ok, in that case I am willing to host an uncompressed file so that no decompression step is necessary. Would it ok if I did that?

**Comment by tupui** [comment]
Things like https://gist.github.com can help to host single plain text files.

**Comment by RPGillespie6** [comment]
Gists have a 50MB limit so that won't work. The repro input is from a CT Scan Slice. The plain text version is going to be 200MB+ so I'll have to host on Google Drive or something. Is that ok?

**Comment by rkern** [comment]
I am not aware that "we" as a project have a policy on the matter. We definitely don't want opaque archives of large amounts of code (but then, it's the "large amounts of code" that's the problem), but we've never had an issue with linking out to data files, like this.

But as a practical matter, the smaller the example data, the better. If it's small enough that we can check it in for our regression tests, that would be perfect.

**Comment by tupui** [comment]
> I am not aware that "we" as a project have a policy on the matter. We definitely don't want opaque archives of large amounts of code (but then, it's the "large amounts of code" that's the problem), but we've never had an issue with linking out to data files, like this.

We don't, but IMHO we should. Since there are exploit to make a reverse shell with an archive, I am not sure we want to bargain until it maybe happens. That would not be a very good press for us. There are more and more attacks all around and being SciPy and having a foot everywhere, I really think we should be careful.

**Comment by RPGillespie6** [comment]
I would like to give a small example data, but like I said, I'm just the guy at my company who said during a standup "What do you mean the same input is giving different results on Linux vs. Windows? That doesn't seem right, can you send me the input and a repro snippet?" and then I raised the issue here. The actual developer who discovered this is a physics researcher who I'm not sure has used GitHub before. So if you want a tiny reproducible example either I'm going to have to ask him to make one and it might take a while or else I'm going to have invest a lot of time to do it myself (and I'm pretty swamped at the moment so I probably will not be able to get around to it). In the interim, I thought it might be useful if you had at least _some_ reproducible input that someone else might be able to distill down.

**Comment by rkern** [comment]
I have confirmed that the file is as described and restored the link.

**Comment by rkern** [comment]
If I had to guess, we're probably running into `long` integer overflow on Windows where it's 32-bit while on Linux it's 64-bit. I don't have a Windows environment handy to test there. What do you get for `ndimage.sum(arr)`? I get `-62666692323` on Linux.

There are a lot of `-1000` values. Are you sure that you wanted those incorporated in the center-of-mass calculations? Negative  mass is allowed, but `-1000` is typically a dummy value in CT images. Doing `center_of_mass(arr, labels=arr >= 0)` might get you what you want in a way that also doesn't trigger the platform difference.

**Comment by rkern** [comment]
Yeah, that's likely the issue, then. If you do want the `-1000`s to count that way, using `arr.astype(np.float32)` would also be a workaround.

**Comment by RPGillespie6** [comment]
Thanks, I'll pass that on! Is the 32-bit/64-bit OS differences thing is a known issue? Seems like `long` should be changed to be platform independent type like `int64_t` (for example). I vote we close this issue and refer to that one since this issue is not specific to `ndimage.center_of_mass` but rather anything that uses `long` to store result under the hood.

**Comment by rkern** [comment]
Something like the following will probably reproduce the platform difference (if not quite so dramatically as to flip signs).

```python
z = np.full((512, 512, 284), -1000, dtype=np.int16)
z[256-16:256+16, 256-16:256+16, 142-16:142+16] = 3000
```

The `long` platform difference is a long-existing "feature" of numpy that is very hard to change at this point at the root (and yes, it _was_ a feature, and these days is more of an albatross, but still hard to change). But it could be worked around here knowing that's a thing. Possibly using `float` accumulators for the summing of the normalizing constant would work since we are casting to `float` elsewhere in the calculation.

**Comment by AnirudhDagar** [comment]
Hey @RPGillespie6 thanks for the report! @rkern, if possible, I'd like to take this issue and send a workaround fix. Thanks for the detailed repro discussion above.
